src/Test/Haskell.idr

```diff
@@ -1,79 +1,159 @@
 module Test.Haskell
 
-import Data.Vect 
-
-import Typedefs
 import Types
+import Typedefs
+
+import Backend
+import Backend.Utils
 import Backend.Haskell
 
+import Text.PrettyPrint.WL
 import Specdris.Spec
 
-%access public export
-
-bit : TDef 0
-bit = TName "Bit" $ TSum [T1, T1]
-
-byte : TDef 0
-byte = TName "Byte" $ pow 8 bit
+import Data.Vect
+import Test
 
-maybe : TDef 1
-maybe = TName "Maybe" $  TSum [T1, TVar 0]
-
-list : TDef 1
-list = TMu "List" [("Nil", T1), ("Cons", TProd [TVar 1, TVar 0])]
+%access public export
 
-maybe2 : TDef 1
-maybe2 = TMu "Maybe2" [("Nothing", T1), ("Just", TVar 1)]
+generate : TDef n -> Doc
+generate = generate {lang=Haskell}
 
-nat : TDef 0
-nat = TMu "Nat" [("Z", T1), ("S", TVar 0)]
+boolForBit : SpecialiseEntry
+boolForBit = MkSpecialiseEntry bit "Bool"
+                               "either (\\ () -> True) (\\ () -> False)"
+                               "\\ x -> if x then Left () else Right ()"
 
-listNat : TDef 0 
-listNat = TMu "ListNat" [("NilN", T1), ("ConsN", TProd [weakenTDef nat 1 LTEZero, TVar 0])]
+charForByte : SpecialiseEntry
+charForByte = MkSpecialiseEntry byte "Char" "undefined" "undefined"
 
-parametricSynonym : TDef 1
-parametricSynonym = TName "ParSyn" maybe
+intForNat : SpecialiseEntry
+intForNat = MkSpecialiseEntry nat "Int"
+                              "id"
+                              "\\ x -> if x >= 0 then x else error \"negative number\""
 
-parametricSynonym2 : TDef 1
-parametricSynonym2 = TName "ParSyn2" maybe2
+generateSpecialised : Vect (S m) SpecialiseEntry -> TDef n -> Doc
+generateSpecialised se td = vsep2 $ map generateCode $ generateDefsSpecialised {lang=Haskell} se _ td
 
 testSuite : IO ()
 testSuite = spec $ do
 
   describe "Haskell code generation tests:" $ do
 
-    it "bit" $ 
-      generate bit 
-        `shouldBe` "\ntype Bit = Either () ()\n"
+    it "bit" $
+      generate bit
+        `shouldBe` text "type" |++| text "Bit" |++| equals |++| text "Either" |++| parens empty |++| parens empty
```

Comment:
A bit nitpicky, but would it make sense to compare the test cases to the actual printed strings (that is,
```
it "bit" $
      (print $ generate bit)
        `shouldBe` "\ntype Bit = Either () ()\n"
```
for instance) so that there is not a bug in the `Doc` rendering?

Reply:
My reasoning for this choice is that whether the tests pass or not shouldn't depend on how we print the `Doc` term. `Doc` needs parameters in order to print (I defined the `print` function here with certain parameters as a convenience), so if one of those is changed, the actual and expected results might differ even though the generated code is essentially the same.

But I agree that the current situation is also pretty clunky, so I don't have a strong preference towards it. Open to arguments and/or alternative suggestions.

Reply:
Yes indeed, harmless reformattings of the same code should ideally not matter... Perhaps we could separately test that `ghc` accepts the code? Since layout matters for eg Haskell, it would be good to know that we actually generate valid code. But indeed, this is a minor issue.